src/IO/StdStreamBufFromReadBuffer.cpp

```diff
@@ -0,0 +1,111 @@
+#include <IO/StdStreamBufFromReadBuffer.h>
+#include <IO/SeekableReadBuffer.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+}
+
+
+StdStreamBufFromReadBuffer::StdStreamBufFromReadBuffer(std::unique_ptr<ReadBuffer> read_buffer_, size_t size_)
+    : read_buffer(std::move(read_buffer_)), seekable_read_buffer(dynamic_cast<SeekableReadBuffer *>(read_buffer.get())), size(size_)
+{
+}
+
+StdStreamBufFromReadBuffer::StdStreamBufFromReadBuffer(ReadBuffer & read_buffer_, size_t size_) : size(size_)
+{
+    if (dynamic_cast<SeekableReadBuffer *>(&read_buffer_))
+    {
+        read_buffer = wrapSeekableReadBufferReference(static_cast<SeekableReadBuffer &>(read_buffer_));
+        seekable_read_buffer = static_cast<SeekableReadBuffer *>(read_buffer.get());
+    }
+    else
+    {
+        read_buffer = wrapReadBufferReference(read_buffer_);
+    }
+}
+
+StdStreamBufFromReadBuffer::~StdStreamBufFromReadBuffer() = default;
+
+int StdStreamBufFromReadBuffer::underflow()
+{
+    char c;
+    if (!read_buffer->peek(c))
+        return std::char_traits<char>::eof();
+    return c;
+}
+
+std::streamsize StdStreamBufFromReadBuffer::showmanyc()
+{
+    return read_buffer->available();
+}
+
+std::streamsize StdStreamBufFromReadBuffer::xsgetn(char_type* s, std::streamsize count)
+{
+    return read_buffer->read(s, count);
+}
+
+std::streampos StdStreamBufFromReadBuffer::seekoff(std::streamoff off, std::ios_base::seekdir dir, std::ios_base::openmode which)
+{
+    if (dir == std::ios_base::beg)
+        return seekpos(off, which);
+    else if (dir == std::ios_base::cur)
+        return seekpos(getCurrentPosition() + off, which);
+    else if (dir == std::ios_base::end)
+        return seekpos(size, which);
```

Comment:
You are ignoring arg off in that case. It might be wrong.
S3 sdk always do call seekg(0, end). However it is legal to call for example seekg(-10, end).

Reply:
Done, I made it correct